### ISSUE TITLE
BSD Firewall module and ipfw's rules format

### ISSUE THREAD

**Issue by crlmahlberg** [webmin/webmin#1425]
I run FreeBSD 12.1 and use _ipfw_ as my firewall. Recently Webmin's _BSD Firewall_ module started working again, after upgrading to version 1.970 (cf. https://github.com/webmin/webmin/issues/1327). That is great. However, now there is another problem: the module seems to use another format for the firewall rules than _ipfw_ itself. Here's an example: the default firewall rules file created by the module is written like this:

```
flush
# Allow all traffic
add 00100 allow all from any to any
```
When _ipfw_ is started from the command line or at boot, with that very `firewall_script` set, I get those error messages:

> /etc/ipfw.rules: flush: not found
> /etc/ipfw.rules: add: not found

That is certainly because the _ipfw.rules_ file is missing `ipfw -q` at the beginning of each rule, which is needed according to the documentation. The rules file is supposed to be written like a shell script. When I adjust the rules file accordingly however, the module does not read out the firewall rules correctly anymore. The rules table on the module's summary page is completely empty. Could this incompatibility be fixed, or is there another reason why the module does not work yet on my server? Thanks for your help!

**Comment by jcameron** [comment]
How are you running this IPFW config file from the command line? The way Webmin executes it is to run `ipfw /etc/ipfw.rules`

**Comment by crlmahlberg** [comment]
I run _ipfw_ as a service (`sudo service ipfw`…). Here's the part from _rc.conf_:
```
firewall_enable="YES"
firewall_script="/etc/ipfw.rules"
firewall_logging="YES"
firewall_quiet="YES"
```
When I do `ipfw /etc/ipfw.rules` directly from the command line, I get
> Line 2: bad command `cmd="ipfw'

Not sure what that means.

**Comment by jcameron** [comment]
I took a closer look at this, and it seems like the problem is that Webmin is uses the wrong format for that config file on FreeBSD.

Is the rc.conf file on your system in /etc ?

**Comment by crlmahlberg** [comment]
Yes, it's _/etc/rc.conf_.

**Comment by jcameron** [comment]
If you empty out the /etc/ipfw.rules file and create new rules, do they get added in a different (hopefully correct) format?

**Comment by crlmahlberg** [comment]
No. I empties out the file, created a new rule through Webmin, and here is the updated rules file:

> add 00100 allow all from any to any

**Comment by bsiege** [comment]
my /etc/ipfw.rules  i made long time ago to work with webmin FreeBSD:

```
#!/bin/sh
# add firewallrules from webmin

ipfw -q -f flush       # Delete all rules
cmd="ipfw -q add "  

while read line; do
	if  echo "$line" | grep  -q -v -e"^#" -e"65535"  ; then	
		rulex="$cmd $line"
		#echo "$rulex"
		$rulex; 	
	fi
done < /usr/local/etc/webmin/ipfw/ipfw.rules
```

Remembers me, that it flushes all existing connections and reload new rules. 
e.g. works but much room for improvement

**Comment by jcameron** [comment]
@crlmahlberg does that config work if you add `ipfw` at the start of each line?

**Comment by crlmahlberg** [comment]
@jcameron No, it does not. @bsiege's rules file causes an error when loaded into _ipfw_ from the command line:

> ipfw: invalid action flush
> ipfw: invalid action add

When I add `ipfw` at the start of each line in _/usr/local/etc/webmin/ipfw/ipfw.rules_, I get the following error:

> ipfw: invalid action ipfw

**Comment by bsiege** [comment]
Had to verify, but for me this script it is still working well on 12.2-STABLE

**Comment by bsiege** [comment]
from man:

> If the world and the kernel get out of sync the ipfw ABI may break,
     preventing you from being able to add any rules.  This can adversely
     affect the booting process.  You can use ipfw disable firewall to
     temporarily disable the firewall to regain access to the network,
     allowing you to fix the problem.

Maybe hitting this?

**Comment by jcameron** [comment]
Looks like Webmin needs some updates to handle the latest IPFW format properly .. I'll take a look.

**Comment by crlmahlberg** [comment]
Hey guys, I'm just curious, are there any updates on this issue? `ipfw disable firewall` doesn't do the trick, or maybe I just don't see it…

**Comment by crlmahlberg** [comment]
Hi there, any updates?

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
